subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+#else
+	const char *name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+	bool hidden;
+	/** Preset current name awareness */
+	ATOMIC_DEFINE(is_client_name_aware, BT_HAS_MAX_CLIENT);
+} has_preset_list[CONFIG_BT_HAS_PRESET_CNT];
+
+/* Separate Preset Control Point configuration for each bonded device */
+static struct has_client {
+	/** Connection object */
+	struct bt_conn *conn;
+	/** Pending preset changed operation per each preset */
+	struct {
+		/* Preset changed pending if flag set */
+		ATOMIC_DEFINE(pending, CONFIG_BT_HAS_PRESET_CNT);
+		/* ChangeId to be sent */
+		uint8_t change_id[CONFIG_BT_HAS_PRESET_CNT];
+	} preset_changed;
+	/** Read preset operation data */
+	struct {
+		/** Next preset to be sent */
+		struct has_preset *preset_pending;
+		/** Number of remaining presets to send */
+		uint8_t num_presets;
+	} read_preset_rsp;
+	union {
+		/** Control Point Indicate parameters */
+		struct bt_gatt_indicate_params ind;
+		/** Control Point Notify parameters */
+		struct bt_gatt_notify_params nfy;
+	} cp_tx_params;
+	/* Control Point Tx work */
+	struct k_work_delayable cp_tx_work;
+	/* Control Point Tx sync state */
+	struct k_work_sync cp_tx_sync;
+	/* Internal flags */
+	ATOMIC_DEFINE(flags, CLIENT_NUM_FLAGS);
+} has_client_list[BT_HAS_MAX_CLIENT];
+
+static struct bt_has has_local;
+static struct bt_has_preset_ops *preset_ops;
+static uint8_t last_preset_id;
+static struct k_work active_preset_work;
+NET_BUF_SIMPLE_DEFINE_STATIC(cp_tx_buf, BT_HAS_CP_TX_MAX_LEN);
+
+static uint8_t client_index(struct has_client *client)
+{
+	ptrdiff_t index = client - has_client_list;
+
+	__ASSERT(index >= 0 && index < ARRAY_SIZE(has_client_list), "Invalid client pointer");
+
+	return (uint8_t)index;
+}
+
+static void client_free(struct has_client *client)
+{
+	BT_DBG("%p", client);
+
+	for (int i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+		struct has_preset *preset = &has_preset_list[i];
+
+		atomic_clear_bit(preset->is_client_name_aware, client_index(client));
+	}
+
+	/* Cancel ongoing work. Since the client can be re-used after this
+	 * we need to sync to make sure that the kernel does not have it
+	 * in its queue anymore.
+	 */
+	k_work_cancel_delayable_sync(&client->cp_tx_work, &client->cp_tx_sync);
+
+	bt_conn_unref(client->conn);
+
+	memset(client, 0, sizeof(*client));
+}
+
+static struct has_client *client_find(struct bt_conn *conn)
+{
+	for (int i = 0; i < ARRAY_SIZE(has_client_list); i++) {
+		if (conn == has_client_list[i].conn) {
+			return &has_client_list[i];
+		}
+	}
+
+	return NULL;
+}
+
+static ssize_t control_point_rx(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *data, uint16_t len, uint16_t offset, uint8_t flags);
+
+static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			     uint16_t len, uint16_t offset)
+{
+	BT_DBG("conn %p attr %p offset %d", conn, attr, offset);
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_LONG);
```

Comment:
Why that error code here, and "BT_ATT_ERR_INVALID_OFFSET" elswhere?

Reply:
Fixed to use Invalid Offset error, however, Core 5.3 Vol 3, Part F, 3.4.4.5 says
"If the attribute value has a fixed length that is less than or equal to
(ATT_MTU - 1) octets in length, then an ATT_ERROR_RSP PDU may be sent
with the Error Code parameter set to Attribute Not Long (0x0B)."

So the behavior was correct